File: app/src/main/java/cn/ifafu/ifafu/data/http/parser/ExamParser.java
package cn.ifafu.ifafu.data.http.parser;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.select.Elements;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collections;
import java.util.List;

import cn.ifafu.ifafu.data.entity.Exam;
import cn.ifafu.ifafu.data.entity.Response;
import cn.ifafu.ifafu.data.entity.User;
import cn.ifafu.ifafu.util.RegexUtils;
import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.ObservableTransformer;

public class ExamParser implements ObservableTransformer<String, Response<List<Exam>>> {

    private String account;
    private int schoolCode;

    public ExamParser(User user) {
        this.account = user.getAccount();
        this.schoolCode = user.getSchoolCode();
    }

    private List<Exam> parse(String html) {
        Document document = Jsoup.parse(html);
        Elements elementsTemp = document.select("table[id=\"DataGrid1\"]");
        if (elementsTemp.size() == 0) {
            return Collections.emptyList();
        }
        Elements elements = elementsTemp.get(0).getElementsByTag("tr");
        List<Exam> list = new ArrayList<>();
        Elements termAndYear = document.select("option[selected=\"selected\"]");
        String year = termAndYear.get(0).text();
        String term = termAndYear.get(1).text();
        for (int i = 1; i < elements.size(); i++) {
            try {
                Exam exam = getExam(elements.get(i).children());
                exam.setTerm(term);
                exam.setYear(year);
                list.add(exam);
            } catch (Exception ignored) {
            }
        }
        return list;
    }

    private Exam getExam(Elements e) {
        List<Integer> numbers = RegexUtils.getNumbers(e.get(3).text());
        Calendar calendar = Calendar.getInstance();
        calendar.set(Calendar.SECOND, 0);
        calendar.set(Calendar.MILLISECOND, 0);

        calendar.set(Calendar.YEAR, numbers.get(0));
        calendar.set(Calendar.MONTH, numbers.get(1) - 1);
        calendar.set(Calendar.DAY_OF_MONTH, numbers.get(2));
        calendar.set(Calendar.HOUR_OF_DAY, numbers.get(3));
        calendar.set(Calendar.MINUTE, numbers.get(4));
        long start = calendar.getTime().getTime();

        calendar.set(Calendar.HOUR_OF_DAY, numbers.get(5));
        calendar.set(Calendar.MINUTE, numbers.get(6));
        long end = calendar.getTime().getTime();

        Exam exam = new Exam();
        exam.setId((long) e.get(0).text().hashCode());
        exam.setName(e.get(1).text());
        exam.setAddress(e.get(4).text());
        exam.setStartTime(start);
        exam.setEndTime(end);
        exam.setSeatNumber(e.get(6).text());
        exam.setAccount(account);

        return exam;
    }

    @Override
    public ObservableSource<Response<List<Exam>>> apply(Observable<String> upstream) {
        return upstream.map(html -> {
            List<Exam> list = parse(html);
            return Response.success(list);
        });
    }
}

File: app/src/main/java/cn/ifafu/ifafu/data/http/parser/ExamParser.kt
package cn.ifafu.ifafu.data.http.parser

import cn.ifafu.ifafu.data.entity.Exam
import cn.ifafu.ifafu.data.entity.Response
import cn.ifafu.ifafu.data.entity.User
import cn.ifafu.ifafu.util.RegexUtils
import org.jsoup.Jsoup
import org.jsoup.select.Elements
import java.util.*
import kotlin.collections.ArrayList

class ExamParser(user: User) : BaseParser<Response<MutableList<Exam>>>() {

    private val account: String = user.account
    private val schoolCode: Int = user.schoolCode

    override fun parse(html: String): Response<MutableList<Exam>> {
        val document = Jsoup.parse(html)
        val elementsTemp = document.select("table[id=\"DataGrid1\"]")
        if (elementsTemp.size == 0) {
            return Response.success(ArrayList())
        }
        val elements = elementsTemp[0].getElementsByTag("tr")
        val list = ArrayList<Exam>()
        val termAndYear = document.select("option[selected=\"selected\"]")
        val year = termAndYear[0].text()
        val term = termAndYear[1].text()
        for (i in 1 until elements.size) {
            val exam = getExam(elements[i].children())
            exam.term = term
            exam.year = year
            list.add(exam)
        }
        return Response.success(list)
    }

    private fun getExam(e: Elements): Exam {
        val exam = Exam()


        val numbers = RegexUtils.getNumbers(e[3].text())
        val calendar = Calendar.getInstance().apply {
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }

        if (numbers.size > 6) {
            calendar.set(Calendar.YEAR, numbers[0])
            calendar.set(Calendar.MONTH, numbers[1] - 1)
            calendar.set(Calendar.DAY_OF_MONTH, numbers[2])

            val start = calendar.apply {
                set(Calendar.HOUR_OF_DAY, numbers[3])
                set(Calendar.MINUTE, numbers[4])
            }.time.time

            val end = calendar.apply {
                set(Calendar.HOUR_OF_DAY, numbers[5])
                set(Calendar.MINUTE, numbers[6])
            }.time.time

            exam.startTime = start
            exam.endTime = end
        }


        exam.id = e[0].text().hashCode().toLong()
        exam.name = e[1].text()
        exam.address = if (e.size > 4) e[4].text() else ""
        exam.seatNumber = if (e.size > 6) e[6].text().run {
            if (this.matches("[0-9]+".toRegex())) this + "号"
            else this
        } else ""
        exam.account = account

        return exam
    }

}
